R/lehdr.R

```diff
@@ -48,67 +51,105 @@ grab_lodes <- function(state, year, lodes_type = c("od", "rac", "wac"),
+    }
+  }
+  
+  # On URL error, the likely culprit is the lack of state/year combination ...
   httr::stop_for_status(httr::HEAD(url),
-                        "data for this combination of features was not found on LODES")
+    paste("retrieve data for this combination of state and year on LODES.",
+          "Please see the most recent LEHD Technical Document for a list of available state/year.",
+          "https://lehd.ces.census.gov/data/lodes/LODES7/"
+    )
+  )
   
+  # Set download directory, check for cache
   download_dir <- path.expand(download_dir)
   if (!dir.exists(download_dir))
     dir.create(download_dir)
   fil <- file.path(download_dir, basename(url))
   
+  # If there is a cache, use it
   if(file.exists(fil)) {
     message(glue::glue("Cached version of file found in {fil}\n Reading now..."))
   } else {
     message(glue::glue("Downloading {url} to {fil} now..."))
     res <- httr::GET(url, httr::write_disk(fil))
   }
   
-  col_types <- cols(w_geocode = col_character(), 
-                    h_geocode = col_character(),
-                    createdate = col_character())
-  
-  if (lodes_type == "rac")
-    col_types$cols$w_geocode <- NULL
-  if (lodes_type == "wac")
-    col_types$cols$h_geocode <- NULL
-  
-  df <- suppressMessages(readr::read_csv(fil, col_types = col_types))
+  # Read in the data
+  lehdr_df <- suppressMessages(readr::read_csv(fil, col_types = col_types))
 
-  df <- df %>% mutate(
+  # Mutate the data based on year and state
+  lehdr_df <- lehdr_df %>% mutate(
     year=year,
-    State=toupper(state))
+    state=toupper(state))
   
-  if (tract==TRUE) {
-    ## Convert *_geocode column(s) to tract id
-    df <- df %>% 
-      mutate_at(vars(ends_with("_geocode")), funs(stringr::str_sub(., 1, 11))) %>% 
-      rename_at(vars(ends_with("_geocode")), 
-                funs(stringr::str_replace(., "geocode", "tract_id")))
-    
-    ## Group by column(s) ending with "_tract_id"
-    tract_cols <- stringr::str_extract(names(df), ".*_tract_id$") %>% na.omit()
-    group_cols <- c("year", "State", tract_cols)
-    group_syms <- rlang::syms(group_cols)
-    df <- df %>% 
-      group_by(!!!group_syms) %>% 
-      summarise_if(is.numeric, funs(sum))
+  # Group by geography, if the agg_geo_to variable is "block" or NULL (not passed via args), 
+  # then skip this step as the data is already in block format
+  if(agg_geo_to != "block" && !is.null(agg_geo_to)) { 
+    id_len <- switch(agg_geo_to, 
+                     "bg" = 12, 
+                     "tract" = 11, 
+                     "county" = 5)
+    lehdr_df <- aggregate_lodes_df(lehdr_df, id_len, agg_geo_to)
   }
 
-  return(df)
+  return(lehdr_df)
+}
+
+#' 
+#' Aggregate data to a certain level dictated by inputs. Internal function.
+#' 
+#' @param geoid_to The number of characters to do the aggregation on from the geoid
+#' @param aggname What the level is called, like "tract" etc
+
+aggregate_lodes_df <- function(.lehdr_df, geoid_to, aggname) {
```

Comment:
Here's the abstracted function for aggregation. I did no export this so it is not public outside the package.